### PR TITLE
8310874: Runthese30m crashes with klass should be in the placeholders during verification

### PR DESCRIPTION
**Pull request by coleenp**
For non-parallel capable class loaders, the class is not put in the placeholders table, so a verification at the wrong time will hit this assert. This change removes the invalid assert.
While studying this code, we also noticed that if an error occurs during class loading after the klass is put in the LoaderConstraints table but before adding to the dictionary, we weren't removing the klass from the LoaderConstraints table.  This change also fixes this.
Tested with inserting a LoaderConstraintTable::verify while running jck tests.  Also tested with failure insertion after the addClass() call to throw OutOfMemoryError.  Also tested with tier1-4.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310874](https://bugs.openjdk.org/browse/JDK-8310874): Runthese30m crashes with klass should be in the placeholders during verification (**Bug** - P3)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15752/head:pull/15752` \
`$ git checkout pull/15752`

Update a local copy of the PR: \
`$ git checkout pull/15752` \
`$ git pull https://git.openjdk.org/jdk.git pull/15752/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15752`

View PR using the GUI difftool: \
`$ git pr show -t 15752`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15752.diff">https://git.openjdk.org/jdk/pull/15752.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15752#issuecomment-1719992949)

### CODE REVIEW

File: src/hotspot/share/classfile/loaderConstraints.cpp [hunk 1]
```diff
@@ -447,6 +443,18 @@ InstanceKlass* LoaderConstraintTable::find_constrained_klass(Symbol* name,
...
 }
 
+// Removes a class that was added to the table then class loading subsequently failed for this class.
+void LoaderConstraintTable::remove_failed_loaded_klass(InstanceKlass* klass,
+                                                       ClassLoaderData* loader) {
+
+  Symbol* name = klass->name();
+  LoaderConstraint *p = find_loader_constraint(name, loader);
+  if (p != nullptr && p->klass() != nullptr && p->klass() == klass) {
+    log_info(class, loader, constraints)("removing klass %s failed to load", name->as_C_string());
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/loaderConstraints.cpp:453]: Suggestion, either:
```
"removing klass %s: failed to load"
```
or
```
"removing klass %s as it failed to load"
```

File: src/hotspot/share/classfile/loaderConstraints.cpp [hunk 2]
```diff
@@ -447,6 +443,18 @@ InstanceKlass* LoaderConstraintTable::find_constrained_klass(Symbol* name,
   return nullptr;
 }
 
+// Removes a class that was added to the table then class loading subsequently failed for this class.
+void LoaderConstraintTable::remove_failed_loaded_klass(InstanceKlass* klass,
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/loaderConstraints.cpp:447]: This only seems to null out the klass ptr not remove the constraint. ??

File: src/hotspot/share/classfile/systemDictionary.cpp [hunk 3]
```diff
@@ -1526,6 +1526,10 @@ InstanceKlass* SystemDictionary::find_or_define_instance_class(Symbol* class_nam
   } else if (HAS_PENDING_EXCEPTION) {
     assert(defined_k == nullptr, "Should not have a klass if there's an exception");
     k->class_loader_data()->add_to_deallocate_list(k);
+
+    // Also remove this InstanceKlass from the LoaderConstraintTable if added.
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionary.cpp:1530]: Why do we not do this in the non-exception case? If k wasa duplicate case then there should be a constraint entry under `defined_k`, so any entry under `k` should also be removed shouldn't it?

**Review by dholmes-ora — COMMENTED** [review]
Simply removing the assertion rather than trying to always use the PHT seems a reasonable simplication.

I'm not really following the remove-on-error part for the constraint table though. It is isn't clear to me what order things are happening in with regard to the loader constraint table addition, the loading of the class (we managed to load it as we have the `instanceKlass` for it) and the subsequent determination that we shouldn't actually define it.

Thanks.

File: src/hotspot/share/classfile/loaderConstraints.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/loaderConstraints.cpp:453]: Actually "failed to load" seems rather vague here, the actual condition is much more specific isn't it?

Comment (coleenp) [reply src/hotspot/share/classfile/loaderConstraints.cpp:453]: The actual condition is in the pending exception and only in the case of OOM will we actually remove the entry.  I know this but the code is more general.

Comment (coleenp) [reply src/hotspot/share/classfile/loaderConstraints.cpp:453]: I'll update the comments.

**Review by coleenp — COMMENTED** [review]
Hope these comments make sense.

File: src/hotspot/share/classfile/loaderConstraints.cpp [hunk 4]
```diff
@@ -450,7 +450,11 @@ void LoaderConstraintTable::remove_failed_loaded_klass(InstanceKlass* klass,
   Symbol* name = klass->name();
   LoaderConstraint *p = find_loader_constraint(name, loader);
   if (p != nullptr && p->klass() != nullptr && p->klass() == klass) {
-    log_info(class, loader, constraints)("removing klass %s failed to load", name->as_C_string());
+    // If this is the klass in the constraint, the error was OOM from the ClassLoader.addClass() call.
+    // other errors during loading will not have added this klass.
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/loaderConstraints.cpp:454]: Nit: Capital O on other

File: src/hotspot/share/classfile/loaderConstraints.cpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/loaderConstraints.cpp:447]: I don't understand what nulling out the klass actually achieves. ?? It seems to me to be allowing for a future attempt to re-load the same named class in the same loader, but IIRC that should not be allowed - at least in terms of resolution errors.

File: src/hotspot/share/classfile/systemDictionary.cpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/systemDictionary.cpp:1530]: Okay

**Review by dholmes-ora — COMMENTED** [review]
Thanks for the explanations and updated comments. I'm still a little vague on the constraint table details though.

**Review by iklam — APPROVED** [review]
Looks good to me. Mulling out the class basically reverts the constraints table to its original state (which says loaders A and B must load the same class for the name N, but neither loader has successfully loaded such a class yet). Here, either A or B attempted to load such a class but that failed due to OOM.

**Review by dholmes-ora — APPROVED** [review]
Okay thanks for the explanations - dangling pointers is much clearer.

**Pull request closed without merging**